### ISSUE TITLE
vader fails in optimized builds on master due to gcc C11 bug

### ISSUE THREAD

**Issue by hppritcha** [open-mpi/ompi#6655]
Looks like somethings going wrong in vader.  I built OMPI master using GNU 7.3 and when
I try to run IMB on a single node using self and vader BTL's I'm getting this:
```
hpp@nid00014:~/imb/src>mpirun -np 8 ./IMB-MPI1
#------------------------------------------------------------
#    Intel (R) MPI Benchmarks 4.1, MPI-1 part    
#------------------------------------------------------------
# Date                  : Mon May 13 16:41:40 2019
# Machine               : x86_64
# System                : Linux
# Release               : 4.4.103-6.38_4.0.153-cray_ari_c
# Version               : #1 SMP Thu Nov 1 16:05:05 UTC 2018 (6ef8fef)
# MPI Version           : 3.1
# MPI Thread Environment: 

# New default behavior from Version 3.2 on:

# the number of iterations per message size is cut down 
# dynamically when a certain run time (per message size sample) 
# is expected to be exceeded. Time limit is defined by variable 
# "SECS_PER_SAMPLE" (=> IMB_settings.h) 
# or through the flag => -time 
  


# Calling sequence was: 

# ./IMB-MPI1

# Minimum message length in bytes:   0
# Maximum message length in bytes:   4194304
#
# MPI_Datatype                   :   MPI_BYTE 
# MPI_Datatype for reductions    :   MPI_FLOAT
# MPI_Op                         :   MPI_SUM  
#
#

# List of Benchmarks to run:

# PingPong
# PingPing
# Sendrecv
# Exchange
# Allreduce
# Reduce
# Reduce_scatter
# Allgather
# Allgatherv
# Gather
# Gatherv
# Scatter
# Scatterv
# Alltoall
# Alltoallv
# Bcast
# Barrier
[nid00014:05737] *** An error occurred in MPI_Bcast
[nid00014:05737] *** reported by process [1620180993,46909632806913]
[nid00014:05737] *** on communicator MPI_COMM_WORLD
[nid00014:05737] *** MPI_ERR_TRUNCATE: message truncated
[nid00014:05737] *** MPI_ERRORS_ARE_FATAL (processes in this communicator will now abort,
[nid00014:05737] ***    and potentially your MPI job)
[nid00014:05709] 2 more processes have sent help message help-mpi-errors.txt / mpi_errors_are_fatal
[nid00014:05709] Set MCA parameter "orte_base_help_aggregate" to 0 to see all help / error messages
```
The problem goes away if I build using ```--enable-debug```.  Turning off vader fast boxes doesn't seem to help.  Things work nominally on v4.0.x so it appears to be something specific to master.

**Comment by ggouaillardet** [comment]
@hppritcha FWIW, I am unable to reproduce the issue in my environment.

are you using a single copy mechanism (such as `xpmem`) and if yes, which one ? what happens if you disable it ?
I cannot download IMB 4.1, can you try with a recent one you can download from https://github.com/intel/mpi-benchmarks ?

**Comment by jsquyres** [comment]
@hppritcha Confirms on the webex today that they did try the different single copy mechanisms and still saw the same behavior.

We're assuming that the broadcast that is failing is a meta data broadcast at the beginning of IMB (i.e., before the actual benchmarks start).

**Comment by hppritcha** [comment]
@ggouaillardet I just picked IMB.  We are actually seeing the problem in MPI_Comm_dup in an application.

**Comment by bosilca** [comment]
Indeed, reproducible as soon as vader is enabled. Turning vader off seems to allow the app to run to completion all the time. I will rename the issue to focus on the real issue.

**Comment by hppritcha** [comment]
git shows that 8291f672 is to blame.
I reduced the problem down to just using osu_latency.  The test hangs in the initial 0 byte ping - pong after a few hundred iterations in the warmup section.

**Comment by hjelmn** [comment]
Oh, fun. Will take a look.

**Comment by hjelmn** [comment]
@hppritcha It might be related to the bug that this works around:

https://github.com/open-mpi/ompi/commit/30119ee339eea086f43e3392352899187a4a73c7#diff-9ce2fbdee8f3b12c6493b3554cc0e991

Wonder if the bug is also impacting C11 atomics. The underlying implementation is the same. Worth trying with --disable-c11-atomics and seeing if the issue goes away.

**Comment by hjelmn** [comment]
Easiest way to verify. Change memory_order_acquire to memory_order_seq_cst here:

https://github.com/open-mpi/ompi/blob/master/opal/include/opal/sys/atomic_stdc.h#L74

**Comment by hjelmn** [comment]
@hppritcha I can confirm that changing the acquire barrier to be sequentially consistent seems to fix the issues. So, this is the same bug as the one we encountered with the __atomic atomic implementation and the same fix is required. I will be sure to add a comment to the effect that once gcc 6.x is no longer supported we can drop the fix.

**Comment by hjelmn** [comment]
Added the appropriate release labels. Will PR the fix shortly.

**Comment by hppritcha** [comment]
closed via #6711.  This problem actually was only on master.
